go/mysql/auth_server.go

```diff
@@ -17,109 +17,233 @@ limitations under the License.
 package mysql
 
 import (
-	"bytes"
 	"crypto/rand"
 	"crypto/rsa"
 	"crypto/sha1"
 	"crypto/sha256"
+	"crypto/subtle"
+	"crypto/x509"
 	"encoding/hex"
 	"net"
-	"strings"
+	"sync"
 
 	"vitess.io/vitess/go/vt/log"
 	"vitess.io/vitess/go/vt/proto/vtrpc"
 	"vitess.io/vitess/go/vt/vterrors"
 )
 
 // AuthServer is the interface that servers must implement to validate
-// users and passwords. It has two modes:
+// users and passwords. It needs to be able to return a list of AuthMethod
+// interfaces which implement the supported auth methods for the server.
+type AuthServer interface {
+	// AuthMethods returns a list of auth methods that are part of this
+	// interface. Building an authentication server usually means
+	// creating AuthMethod instances with the known helpers for the
+	// currently supported AuthMethod implementations.
+	//
+	// When a client connects, the server checks the list of auth methods
+	// available. If an auth method for the requested auth mechanism by the
+	// client is available, it will be used immediately.
+	//
+	// If there is no overlap between the provided auth methods and
+	// the one the client requests, the server will send back an
+	// auth switch request using the first provided AuthMethod in this list.
+	AuthMethods() []AuthMethod
+
+	// Default auth method description that the auth server sends during
+	// the initial server handshake. This needs to be either `mysql_native_password`
+	// or `caching_sha2_password` as those are the only supported auth
+	// methods during the initial handshake.
+	//
+	// It's not needed to also support those methods in the AuthMethods(),
+	// in fact, if you want to only support for example clear text passwords,
+	// you must still return `mysql_native_password` or `caching_sha2_password`
+	// here and the auth switch protocol will be used to switch to clear text.
+	DefaultAuthMethodDescription() AuthMethodDescription
+}
```

Comment:
nit: comment to start with function name `DefaultAuthMethodDescription`

Reply:
Will fix!